homeassistant/helpers/entityfilter.py

```diff
@@ -39,71 +39,122 @@ def generate_filter(include_domains: Iterable[str],
                     exclude_domains: Iterable[str],
                     exclude_entities: Iterable[str]) -> Callable[[str], bool]:
     """Return a function that will filter entities based on the args."""
-    include_d = set(include_domains)
-    include_e = set(include_entities)
-    exclude_d = set(exclude_domains)
-    exclude_e = set(exclude_entities)
-
-    have_exclude = bool(exclude_e or exclude_d)
-    have_include = bool(include_e or include_d)
-
     # Case 1 - no includes or excludes - pass all entities
-    if not have_include and not have_exclude:
-        return lambda entity_id: True
-
-    # Case 2 - includes, no excludes - only include specified entities
-    if have_include and not have_exclude:
-        def entity_filter_2(entity_id: str) -> bool:
-            """Return filter function for case 2."""
+    if (not include_domains and not include_entities
+            and not exclude_domains and not exclude_entities):
+        return EntityFilter(include_all=True)
+
+    return EntityFilter(include_domains=include_domains,
+                        include_entities=include_entities,
+                        exclude_domains=exclude_domains,
+                        exclude_entities=exclude_entities)
+
+
+class EntityFilter:
```

Comment:
I would like to see your benchmark, and especially see how diverse it is.

This method is potentially run a lot of times for every state change, and so adding a bunch of extra calls and lookups will impact it, just not sure by how much.

Reply:
It should be possible to keep the original implementation, and have a new class-based implementation for inside templates that inside the `__call__` just calls `return generate_filter(self.include_entities, …)(entity_id)`

Reply:
The benchmark was running the test case 100,000 times for each test.

I did consider doing that, but you end up creating more garbage because this results in a temporary method object with every call. Really the only difference between the two cases is less than ten extra boolean compares, so it's not that big a deal. 

You could freeze the object, make the sets unmodifiable, and replace `__call__` with the short code path but that's a lot of extra code to avoid a small number of boolean compares. The bulk of the computation is in the string compares and the list access anyhow, and there's no getting around that.

Reply:
alright lets keep it as is.

Reply:
You mean keep it as it is now, or as it was?

Reply:
I had another look at the `__call__` method, it *could* be optimised a bit, add a freeze method to the class / option to the `__init__` and then replace `__call__` with a short path, but I don't really think it's gonna make a whole lot of odds because the short path would only cut out a few Boolean comparisons. If we're really loading that section up enough to make a difference, then there's far worse things to worry about (like the set searches for example, or everything else).

But for the sake of completeness... there's a faster way to rewrite the decision on which algorithm to use too, something like:

```python

algo_map = (
    (self._include_entities and 0x1) |
    (self._exclude_entities and 0x2) |
    (self._include_domains and 0x4) |
    (self._exclude_domains and 0x8))

return self._ALGS[algo_map]()
```

And you load up `_ALGS` statically with your 16 various code paths (some are duplicates). To freeze the filter you replace the existing sets with `frozenset`s, and set `__call__` to be the right algo.
                     
I can do the optimisation if you want, but I think it would really be premature optimisation.

Reply:
keep it as it is now.